multiuser/keycloak/che-multiuser-keycloak-server/src/main/java/org/eclipse/che/multiuser/keycloak/server/KeycloakEnvironmentInitalizationFilter.java

```diff
@@ -73,87 +90,125 @@ public KeycloakEnvironmentInitalizationFilter(
     }
 
-    final HttpSession session = httpRequest.getSession();
-    Subject subject = (Subject) session.getAttribute("che_subject");
-    if (subject == null || !subject.getToken().equals(token)) {
-      Jwt jwtToken = (Jwt) httpRequest.getAttribute("token");
-      if (jwtToken == null) {
-        sendError(response, 401, "Cannot detect or instantiate user.");
-      }
-      Claims claims = (Claims) jwtToken.getBody();
+    Claims claims;
+    try {
+      final HttpSession session = httpRequest.getSession();
+      Subject subject = (Subject) session.getAttribute("che_subject");
+      if (subject == null || !subject.getToken().equals(token)) {
+        Jws<Claims> jwt = jwtParser.parseClaimsJws(token);
+        claims = jwt.getBody();
+        LOG.debug("JWT = ", jwt);
 
-      try {
-        String username =
-            claims.get(
-                keycloakSettings.get().get(KeycloakConstants.USERNAME_CLAIM_SETTING), String.class);
-        if (username == null) { // fallback to unique id promised by spec
-          // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
-          username = claims.getIssuer() + ":" + claims.getSubject();
-        }
-        String email = claims.get("email", String.class);
-        String id = claims.getSubject();
-
-        if (isNullOrEmpty(email)) {
-          boolean userNotFound = false;
-          try {
-            userManager.getById(id);
-          } catch (NotFoundException e) {
-            userNotFound = true;
+        try {
+          String username =
+              claims.get(
+                  keycloakSettings.get().get(KeycloakConstants.USERNAME_CLAIM_SETTING),
+                  String.class);
+          if (username == null) { // fallback to unique id promised by spec
+            // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
+            username = claims.getIssuer() + ":" + claims.getSubject();
           }
-          if (userNotFound) {
+          String email = claims.get("email", String.class);
+          String id = claims.getSubject();
+
+          if (isNullOrEmpty(email)) {
+            boolean userNotFound = false;
             try {
-              EnvironmentContext.getCurrent()
-                  .setSubject(new SubjectImpl(username, id, token, true));
-              Map<String, String> profileAttributes =
-                  keycloakProfileRetriever.retrieveKeycloakAttributes();
-              email = profileAttributes.get("email");
-              if (email == null) {
-                sendError(
-                    response,
-                    400,
-                    "Unable to authenticate user because email address is not set in keycloak profile");
-                return;
+              userManager.getById(id);
+            } catch (NotFoundException e) {
+              userNotFound = true;
+            }
+            if (userNotFound) {
+              try {
+                EnvironmentContext.getCurrent()
```

Comment:
such a long method, please consider refactoring it like in the following way, maybe you're able to extract other parts as well.
```java
...
 if (username == null) { // fallback to unique id promised by spec
            // https://openid.net/specs/openid-connect-basic-1_0.html#ClaimStability
            username = claims.getIssuer() + ":" + claims.getSubject();
          }
          String id = claims.getSubject();
          String email = retrieveEmail(claims, username, id, token);

          if (email == null) {
            sendError(
                response,
                400,
                "Unable to authenticate user because email address is not set in keycloak profile");
            return;
          }

          User user = userManager.getOrCreateUser(id, email, username);
...

  private String retrieveEmail(Claims claims, String username, String id, String token)
      throws ServerException {
    String email = claims.get("email", String.class);
    if (!isNullOrEmpty(email)) {
      return email;
    }

    boolean userNotFound = false;
    try {
      userManager.getById(id);
    } catch (NotFoundException e) {
      userNotFound = true;
    }

    if (!userNotFound) {
      //TODO Is it OK ???
      return null;
    }

      try {
        EnvironmentContext.getCurrent()
            .setSubject(new SubjectImpl(username, id, token, true));
        Map<String, String> profileAttributes =
            keycloakProfileRetriever.retrieveKeycloakAttributes();
        return profileAttributes.get("email");
      } finally {
        EnvironmentContext.reset();
      }
  }
```